pgml-sdks/python/pgml/pgml/database.py

```diff
@@ -0,0 +1,93 @@
+
+import logging
+from rich.logging import RichHandler
+from rich.progress import track
+
+from typing import List, Dict, Optional
+
+from .collection import Collection
+from .dbutils import run_select_statement, run_create_or_insert_statement, run_drop_or_delete_statement
+import os
+
+FORMAT = "%(message)s"
+logging.basicConfig(
+    level=os.environ.get("LOGLEVEL", "ERROR"),
+    format=FORMAT,
+    datefmt="[%X]",
+    handlers=[RichHandler()],
+)
+log = logging.getLogger("rich")
+
+
+class Database:
+    def __init__(self, conninfo: str, min_connections: Optional[int] = 1) -> None:
+        """
+        This function initializes a connection pool and creates a table in a PostgreSQL database if it does
+        not already exist.
+
+        :param conninfo: A string containing the connection information for the PostgreSQL database, such
+        as the host, port, database name, username, and password
+        :type conninfo: str
+        :param min_connections: The minimum number of connections that should be maintained in the
+        connection pool at all times. If there are no available connections in the pool when a new
+        connection is requested, a new connection will be created up to the maximum size of the pool,
+        defaults to 1
+        :type min_connections: Optional[int] (optional)
+        """
+        self.conninfo = conninfo
+        self.pool = ConnectionPool(conninfo, min_size=min_connections)
+        log.info("Creating table pgml.collections")
+        create_statement = "CREATE TABLE IF NOT EXISTS pgml.collections \
+                            (id  serial8 PRIMARY KEY,\
+                            created_at  timestamptz NOT NULL DEFAULT now(),\
+                            name  text NOT NULL)"
+        conn = self.pool.getconn()
+        run_create_or_insert_statement(conn, create_statement)
+        self.pool.putconn(conn)
+
+    def create_collection(self, name: str) -> Collection:
+        """
+        This function creates a new collection in a PostgreSQL database if it does not already exist and
+        returns a Collection object.
+
+        :param name: The name of the collection to be created
+        :type name: str
+        :return: A Collection object is being returned.
+        """
+        # Get collection names
+        conn = self.pool.getconn()
+        results = run_select_statement(conn, "SELECT name FROM pgml.collections")
+        name = name.lower()
+        names = [res["name"] for res in results]
+
+        if name in names:
+            log.info("Collection with name %s already exists.." % name)
+        else:
+            insert_statement = "INSERT INTO pgml.collections (name) VALUES ('%s')" % (
+                name
+            )
+            run_create_or_insert_statement(conn, insert_statement)
+            # Create collection object
+        self.pool.putconn(conn)
+        return Collection(self.pool, name)
+
+    def delete_collection(self, name: str) -> None:
```

Comment:
It is the only way to actually free storage space. Maybe delete should be made two phase?

Reply:
Yeah, we could have some archiving mechanism, which simply alters the schema to `name` + '_archive', and then enable delete or restore on archives.

Reply:
Adding more storage space is cheap. Recovering lost data is expensive and scary.

Reply:
Also, we should support "undo".

Reply:
that's what I meant by "restore"

Reply:
Well, it would be better to leave the schema as is and simply mark it archived in the `collections` table. That makes it easier to query for the condition where it can be deleted: if the `archived` flag is set.

Reply:
That does not guarantee people aren't still using the table, if they are ignoring the flag by using poorly written homegrown queries, which is a potentially desirable outcome for projects that outgrow simple SDK designs, and what to leverage the full expressive power of SQL.

Reply:
In order to access the collection, they look up the collection with the SDK. The SDK looks in the `collections` table for the namespace. The SDK should return an error if the collection is archived.

If they don't look up the namespace with the SDK, how are they going to get the table? People can of course write queries that use the table whether the namespace is moved to `...archived` or not. However, if we have a query builder, that is going to get the namespace and ultimately table for them by first looking in the `collections` table, and the `archived` flag is there.